_posts/2017-08-21-Write-Efficient-Deep-Learning-Operators-with-TVM-A-Depthwise-Convolution-Example.md

```diff
@@ -0,0 +1,412 @@
+        }
+      }
+    }
+  }
+}
+```
+
+As we can see, when `num_vthread_y = 2` and `num_vthread_x = 2`, the 32 x 32 channel is divided into four sub-channels of 16 x 16.
+Each thread computes four output elements at a time, one element in one sub-channel.
+
+Below is the result with Filter = [256, 1, 3, 3], stride = [1, 1], blocking_h = 32, blocking_w = 32:
+
+<style>
+table th:nth-of-type(1) {
+    width: 120px;
+}
+table th:nth-of-type(2) {
+    width: 120px;
+}
+</style>
+
+| Case | Input            | num_thread_y, num_thread_x | num_vthread_y, num_vthread_x | TVM SAME pad (us) |
+|:----:|:----------------:|:--------------------------:|:----------------------------:|:-----------------:|
+| 1    | [1, 256, 96, 96] | 8, 8                       | 1, 1                         | 132.5             |
+| 2    | [1, 256, 96, 96] | 8, 8                       | 1, 4                         | 103.1             |
+| 3    | [1, 256, 96, 96] | 4, 32                      | 1, 1                         | 95.9              |
+| 4    | [1, 256, 96, 96] | 8, 16                      | 1, 2                         | 90.9              |
+
+Case 2 is faster than case 1. It's because in case 2 `num_thread_x=8` and `num_vthread_x=4` together brings coalesced memory access, as illustrated below
+(each color represents the workload of one thread):
+
+{:center}
+![image](/images/depthconv_tutorial/vthread_and_strided_pattern.png){: width="80%"}
+{:center}
+
+In theory case 3 4 should be the same fast, since they have the same workload per thread, and both enjoy coalesced memory access. Somehow case 4 is just a little faster.
+
+Still remember tensorflow's speed? It's 206.0us, and now TVM is 2.3x faster. 387.4 -> 132.5 -> 95.9 -> 90.9, blocking helps the most; tuning thread numbers saves 37us;
+vthread saves additional 5us.
+
+In fact, TVM can be extremely faster than tensorflow with large kernel size or channel_multiplier (because of more filter reuse) :
+
+| Input            | Filter         | stride | TensorFlow SAME pad (us) | TVM SAME pad (us) | How faster is TVM |
+|:----------------:|:--------------:|:------:|:------------------------:|:-----------------:|:-----------------:|
+| [1, 256, 96, 96] | [256, 1, 3, 3] | [1, 1] | 206.0                    | 90.9              | 2.3x              |
+| [1, 256, 96, 96] | [256, 1, 5, 5] | [1, 1] | 597.6                    | 128.9             | 4.6x              |
+| [1, 256, 96, 96] | [256, 2, 3, 3] | [1, 1] | 659.9                    | 143.7             | 4.6x              |
+| [1, 256, 96, 96] | [256, 2, 5, 5] | [1, 1] | 1203.9                   | 170.5             | 7.1x              |
+ 
+## 3. Operator Fusion
+
+TVM can do more than schedule a single operator.
```

Comment:
One typical optimization that we can do in deep learning is cross operator fusion, that computes multiple operators together in a single kernel, without saving back to global memory. TVM support that out of the box.